File: Baekjoon/dongwan/좋은수열.py
N = int(input())

res = 10e9

def is_good(num) :

    l = len(num)
    for idx in range(1,l//2+1) :
        if num[-idx:] == num[-(idx*2):-idx] :
            return False 
    else :
        return True

def dfs(x) :
    global N, res 
    if len(x) == N :
        print(x)
        exit()
    for i in ['1','2','3'] :
        if is_good(x + i) :
            dfs(x+i)
    return
lst = ['1','2','3','4']

dfs('1')

File: Baekjoon/yerim/좋은수열.cpp
// [2661] 좋은수열
#include <iostream>
#include <cstring>
using namespace std;

int n;

void input()
{
    cin >> n;
}

void dfs(string str, char num)
{
    str += num;
 
    bool flag = true;
    int len = str.length();
    
    for(int i = len - 1; i >= len/2; i--) {
        int gap = len - i;
        if(i - gap < 0) break;
        if(str.substr(i, gap) == str.substr(i - gap, gap)) {
            flag = false;
            break;
        }
    }
    if(flag == false) return;

    if(len == n) {
        cout << str << '\n';
        exit(0);
    }

    for(int i = 1; i <= 3; i++) {
        dfs(str, '0' + i);
    }
}

void solve()
{
    if(n == 1) cout << "1" << '\n';
    else {
        for(int i = 1; i <= 3; i++) {
            dfs("1", '0' + i);
        }
    }
}

int main(void)
{
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    input();
    solve();

    return 0;
}